### ISSUE TITLE
Win32 Exception C0000005in DolphinVM7.DLL

### ISSUE THREAD

**Issue by pjavier29** [dolphinsmalltalk/Dolphin#595]
I have a development that consists of 4 applications that connect between them.
These applications during their execution, in a random way, have an instantaneous closing (without accusing error). The error report is generated in the installation folder, they are attached to the task. The extension of the error files is .ERRORS but I changed it to .txt to be able to attach them. The executable files were created with version 7.0.32. I also tried with version 7.0.56 and the same problem occurs.
The applications connect through the Socket2 class to be able to send and receive information between them.
Any suggestion of the possible problem?
Thank you very much and sorry for my bad English.

[CDT_3_02_002.txt](https://github.com/dolphinsmalltalk/Dolphin/files/2236660/CDT_3_02_002.txt)
[Servidor_5_02_002.txt](https://github.com/dolphinsmalltalk/Dolphin/files/2236661/Servidor_5_02_002.txt)
[Usuario_6_02_002.txt](https://github.com/dolphinsmalltalk/Dolphin/files/2236662/Usuario_6_02_002.txt)

**Comment by blairmcg** [comment]
Looks like memory corruption. Check that the buffers you are passing to `Socket2>>receiver:count:startingAt:` are large enough such that there are at least `countInteger` bytes after `startInteger`. 

```
receive: aByteObject count: countInteger startingAt: startInteger 
	| received |
	received := 0.
	[received < countInteger] whileTrue: 
			[received := received + (self 
								receiveSome: aByteObject
								count: countInteger
								startingAt: received + startInteger)].
	^aByteObject
```

Unfortunately with the way that is written, it doesn't have the usual level of safety one would expect in Smalltalk. It could usefully check that `(countInteger  + startInteger - 1) <= aByteObject size` and raise an error if not.

If it's not that (e.g. you are using `Socket2>>receive:` only), I can only suggest running a debug build under the VS debugger to see if some assertion fires. That might at least help to narrow it down. You could also try the 7.1 VM to see if the behaviour is any different as I fixed some overlapped call bugs when trying to bootstrap it. To do so, however, will require loading your code into a new image and fixing any issues associated with UTF-encoded strings. That might be a lot of effort to find you still have the problem, so I wouldn't start there.

**Comment by pjavier29** [comment]
Thank you very much for the answer.

I am using SocketWriteStream to send data and SocketReadStream to receive data. Finally, these classes use the methods "sendSome: aByteObject count: lengthInteger startingAt: startInteger" and 
"receiveSome: aByteObject count: lengthInteger startingAt: startInteger", respectively. 

I added the following validation in both methods: 
	| bytesReceived |

	lengthInteger + startInteger - 1 <= aByteObject size
		ifFalse: [Error signal: 'Problema recepcion  con el socket'].

	bytesReceived := WS2_32Library default
				recv: self asParameter
				buf: aByteObject yourAddress + startInteger - 1
				len: lengthInteger
				flags: 0.
	bytesReceived > 0 ifTrue: [^bytesReceived].
	bytesReceived = 0 ifTrue: [^self errorSocketClosed].
	^self socketErrorOverlap

The problem manifests in the same way. Maybe the problem is not with the sockets. I add the generated error report to the message. Thank you very much for your time.

********************************************************************************
************************** Dolphin Crash Dump Report ***************************

13:55:43 p.m., 30/07/2018: Servidor_5_02_002.exe caused an unhandled Win32 Exception C0000005
at 10009EF0 in module 10000000 (C:\Program Files (x86)\MyApp\Servidor\DolphinVM7.DLL)

*----> Exception Parameters <----*
00000000	
21270704	

*----> CPU Context for thread 0x17b4 <----*
EAX = 00B15208 EBX = 04B3CB88 ECX = 10050010
ESI = 103B4100 EDI = 103B4100 EIP = 10009EF0
ESP = 0019F6B4 EBP = 04B3CB88 EFL = 00010212
CS = 0023 SS = 002B DS = 002B
ES = 002B FS = 0053 GS = 002B

*----> VM Context <----*
Process: {04B3CB88:size 5 words, suspended frame 103ADFD0, priority 136146240, callbacks 136149528
last failure 0:0, FPE mask 134381576, thread 00000000}
Active Method: Process>>suspendingList
IP: 0284BF58 (0)
SP: 0764032C
BP: 07640318 (11275733)
ActiveFrame: {07640318: cf 103ADFD0, sp 00000000, bp 07640318, ip 0, SmallInteger(Process)>>suspendingList}
	receiver: 61997440


New Method: Process>>suspendingList
Message Selector: suspendingList

*----> Stack Back Trace <----*
{07640318: cf 103ADFD0, sp 00000000, bp 07640318, ip 0, SmallInteger(Process)>>suspendingList}
	receiver: 61997440


***Unable to generate complete exception report (c0000005)

***** End of crash report *****

**Comment by pjavier29** [comment]
Based to the advice you gave us, we have optimized our communication class and we have been able to find the source of the problem. The problem is in the creation of processes, when a number is exceeded, the error occurs in the virtual machine. We did not have this problem with Dolphin SmallTalk 5 as it seems to be more robust about it.
Here are some examples to reproduce the problem.

To run on Dolphin SmallTalk 7

* `1 a: 10000 do: [: i | [1000 factorial] fork]` - Close the image whit unhandled Win32 Exception before reaching 2400 proccess.

* `[true] whileTrue: [[1000 factorial] fork]` - Close the image with unhandled Win32 Exception.

To run on Dolphin SmallTalk 5.15

* `1 a: 10000 do: [: i | [1000 factorial] fork]` - Works perfectly.

* `[true] whileTrue: [[1000 factorial] fork]` - Produces a problem in ProcessorScheduler in the "noMemory: interruptArg" method but the image still running.

Is there any way to get the same behavior as in Dolphin SmallTalk 5?

Regards.

**Comment by blairmcg** [comment]
Well done for tracking it down.

I think you are running out of virtual (not physical) memory. Dolphin is a 32-bit process, which means that there is an absolute limit of 4Gb of addressable storage when running on a 64-bit host OS (3Gb on a 32-bit OS). And ironically this is (in theory at least) twice what Dolphin 5 can address, because it was not "large address aware" and was therefore limited to 2Gb.

The reason for the number of processes that you can launch being lower on Dolphin 7 is largely because of the different default allocations for two things:
1. Default stack reservation - 128K slots (or 512Kb) on D7, and only 32K slots (or 128Kb) on D5. This is the maximum size the stack can grow to before you will get a stack overflow exception - the actual allocation is usually much smaller. Although 512Kb is not a lot of memory individually, 2400 such stacks will be reserving 1.2Gb of the available virtual address space. When I run an experiment with your code, but writing `i` to a file, it gets to 5954 before it crashes. That is the equivalent of nearly 3Gb with 128Kb stacks, although it is difficult to know how many of those processes are actually live when it crashes.
1. Default object table reservation - this is much larger in D7, and has been increased over the years to satisfy requests to increase the maximum number of objects. In D7 this is 24*1024*1024 objects (about 25 million). In D5 it was 8 million. Each OT entry is 16 bytes. Although individually this makes a much greater difference, compared to the process stacks it is fairly insignificant - 

One would have to look at a memory map of a running D7 process and compare it with a running D5 process, but D5 contained less and used much simpler UI, so it will also have less memory reserved for various OS services.

The most significant difference, however, is the default maximum stack reservation. I don't recall why or when this was increased, but it really doesn't need to be as large. You can easily tune it down by altering the Process.DefaultMaxStack constant:

```
Process addClassConstant: 'DefaultMaxStack' value: 32*1024
```

I did this and re-ran the experiment, and 10000 was written to the file, but Dolphin still crashed. It may still be exhausting memory and preventing some other virtual allocation succeeding, or it may be some bug. So I tried reducing the maximum OT size as well. This can be done by saving the image and passing a different maximum number of objects to the snapshot primitive:

```
SessionManager current
		primSnapshot: SessionManager current imageFileName
			backup: true
			type: 0
			maxObjects: 8192*1024.
```
Then exit (without saving) and restart. You can inspect `MemoryManager current` to see the before and after effect on the `maxObjects` instance variable.

When I did this, it still crashed, although if I lowered the limit to 7500 it did complete successfully. Mind you I didn't get the same results as you did with an old copy of Dolphin 5.0 I found on my disk. It failed before 10000 too - the maximum seemed somewhere between 8500 and 9000 before it starts reporting out of memory. It doesn't crash though.

Regarding the misbehavior when the limit is reached, that sounds like a bug. I will look into it. In the meantime you could try at least changing the default max stack to see if that improves things for you. It may extend the limit enough to mean you don't get near enough to induce a crash.

Perhaps it goes without saying, but processes are a limited resource, and eventually you will hit that limit. If you are also performing overlapped calls from these processes (which Sockets2 will be doing), then each Smalltalk Process will also be associated with a native thread. This will have its own stack reservation, and you are going to hit the limit very much sooner. I don't know whether Windows will support 10,000 threads in one process, but I doubt it. I don't know enough about your application to know whether you really need so many processes, but it would certainly be worth looking at whether the design could be altered to avoid forking a very large number of processes. For example by pushing some work onto a shared queue (or queues) and processing it that way, instead of forking a process for every work item. Although it might seem like this would reduce your throughput, the reverse might be true. Only a single OS thread (and therefore a single OS core) can be actively processing Smalltalk code at any one time, so there is no real concurrency anyway and context switching overhead will in fact reduce throughput if you have more than a few concurrent Processes running.

Regards

Blair

**Comment by rejurime** [comment]
Just out of curiosity, I've tried the same code in my Dolphin Smalltalk X6.1. I've got the same behavior that @pjavier29 describes in Dolphin Smalltalk 5.1.5. The Error is handled in the image, but in Dolphin 7.0.56 and 7.1.2 produces an image shutdown.

![captura](https://user-images.githubusercontent.com/7131403/45336275-3cb4d180-b559-11e8-9a3c-29aede63a317.png)

The errors file looks like this
``` smalltalk
00:20:02, martes, 11 de septiembre de 2018: Unhandled exception - an OutOfMemoryError('No hay suficientes recursos de memoria disponibles para completar esta operación. (16rE: No hay suficientes recursos de memoria disponibles para completar esta operación.)')

ProcessorScheduler>>noMemory:
[] in ProcessorScheduler>>vmi:list:no:with:
BlockClosure>>ifCurtailed:
ProcessorScheduler>>vmi:list:no:with:
SmallInteger(Object)>>doesNotUnderstand:
SmallInteger(Behavior)>>new:max:
Process class>>new
Process class>>forContext:priority:
BlockClosure>>newProcess
BlockClosure>>forkAt:
InputState>>forkMain
ProcessorScheduler>>forkMain
ProcessorScheduler>>forkMainIfMain
WalkbackDialog(DialogView)>>runModalLoop
WalkbackDialog>>runModalLoop
[] in WalkbackDialog(DialogView)>>showModalTo:
BlockClosure>>ifCurtailed:
BlockClosure>>ensure:
WalkbackDialog(DialogView)>>showModalTo:
WalkbackDialog(DialogView)>>showModal
----------------------------------------------------------------------------------------------------

00:52:53, martes, 11 de septiembre de 2018: Unhandled exception - an OutOfMemoryError('Unrecognised HRESULT - 16r8007000E (16rE: Unrecognised HRESULT - 16r8007000E)')

ProcessorScheduler>>noMemory:
[] in ProcessorScheduler>>vmi:list:no:with:
BlockClosure>>ifCurtailed:
ProcessorScheduler>>vmi:list:no:with:
SmallInteger(Object)>>doesNotUnderstand:
SmallInteger(Behavior)>>new:max:
Process class>>new
Process class>>forContext:priority:
BlockClosure>>newProcess
BlockClosure>>fork
UndefinedObject>>{unbound}doIt
CompiledExpression>>value:
SmalltalkWorkspace>>evaluateRange:ifFail:debug:
SmalltalkWorkspace>>evaluateItIfFail:debug:
SmalltalkWorkspace>>evaluateItIfFail:
SmalltalkWorkspace>>evaluateIt
Symbol>>forwardTo:
CommandDescription>>performAgainst:
[] in Command>>value
BlockClosure>>ifCurtailed:
----------------------------------------------------------------------------------------------------
```

**Comment by blairmcg** [comment]
That is more useful info to help narrow it down, thanks

**Comment by blairmcg** [comment]
The issue was introduced in VM version v7.0.16. It was not present in v7.0.15, which behaves as earlier Dolphin versions and shows an out of memory walkback.

**Comment by blairmcg** [comment]
And it turns out there was very little change between these versions. Perhaps the most like candidate is: 

e9d8da0a42ccb4af849fdbc30e482c00c6575c11 DolphinVM should be built "large address aware" #73

How ironic.

So @pjavier29, try these steps in 7.0.56 or 7.1.2:
1. Open the dolphin.targets file in VisualStudio and comment out the `<LargeAddressAware>true</LargeAddressAware>` setting (or change it to false).
1. Rebuild in VS
1. Start Dolphin
1. [Optionally] Execute the `Process addClassConstant: 'DefaultMaxStack' value: 32768` expression to reduce the default max stack reservation
1. Rerun your test expression

You should find you get similar behavior and the out-of-memory error will result in a walkback and not crash the image.

**Comment by pjavier29** [comment]
I have tested the recommendations on version 7.0.56 and it worked. Now we can capture the error in the image and handle it. I'm going to migrate to this version.
On the other hand, I could trace the problem in our code that generated the memory overflow, we used a fork to process each incoming message instead of putting them in a queue (as you recommended), so we are also working on solving that. Along with the improved version 7.0.56 I think we will have good results.
Thank you very much for your time.
Regards.

**Comment by blairmcg** [comment]
This is fixed in 7.0 and master branches now. It doesn't actually require any VM changes to fix, but I made some minor alterations anyway to help with debugging, so I'll release new versions.
Note that there is still a related issue with `ProcessCopier` when process addresses cross the 2Gb boundary, but this is probably only relevant if using Continuations. I'll fix this separately as a new issue.
